sdkbuild/typescript.go

```diff
@@ -0,0 +1,226 @@
+
+// BuildTypeScriptProgram builds a TypeScript program. If completed
+// successfully, this can be stored and re-obtained via
+// TypeScriptProgramFromDir() with the Dir() value (but the entire BaseDir must
+// be present too).
+func BuildTypeScriptProgram(ctx context.Context, options BuildTypeScriptProgramOptions) (*TypeScriptProgram, error) {
+	if options.BaseDir == "" {
+		return nil, fmt.Errorf("base dir required")
+	} else if options.Version == "" {
+		return nil, fmt.Errorf("version required")
+	} else if len(options.HarnessPaths) == 0 {
+		return nil, fmt.Errorf("at least one harness path required")
+	}
+
+	// Create temp dir if needed that we will remove if creating is unsuccessful
+	success := false
+	var dir string
+	if options.DirName != "" {
+		dir = filepath.Join(options.BaseDir, options.DirName)
+	} else {
+		var err error
+		dir, err = os.MkdirTemp(options.BaseDir, "program-")
+		if err != nil {
+			return nil, fmt.Errorf("failed making temp dir: %w", err)
+		}
+		defer func() {
+			if !success {
+				// Intentionally swallow error
+				_ = os.RemoveAll(dir)
+			}
+		}()
+	}
+
+	// Create package JSON
+	var packageJSONDepStr string
+	if strings.ContainsAny(options.Version, `/\`) {
+		if _, err := os.Stat(filepath.Join(options.Version, "package.json")); err != nil {
+			return nil, fmt.Errorf("failed finding package.json in version dir: %w", err)
+		}
+
+		// Have to build the local repo
+		if st, err := os.Stat(filepath.Join(options.Version, "node_modules")); err != nil || !st.IsDir() {
+			// Only install dependencies, avoid triggerring any post install build scripts
+			cmd := exec.CommandContext(ctx, "npm", "ci", "--ignore-scripts")
+			cmd.Dir = options.Version
+			cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
+			if err := cmd.Run(); err != nil {
+				return nil, fmt.Errorf("failed installing SDK deps: %w", err)
+			}
+
+			// Build the SDK, ignore the unused `create` package as a mostly insignificant micro optimisation.
+			cmd = exec.CommandContext(ctx, "npm", "run", "build", "--", "--ignore", "@temporalio/create")
+			cmd.Dir = options.Version
+			cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
+			if err := cmd.Run(); err != nil {
+				return nil, fmt.Errorf("failed building SDK: %w", err)
+			}
+		}
+
+		// Create package.json updates
+		localPath, err := filepath.Abs(options.Version)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get absolute path from version path: %w", err)
+		}
+		pkgs := []string{"activity", "client", "common", "internal-workflow-common",
+			"internal-non-workflow-common", "proto", "worker", "workflow"}
+		for _, pkg := range pkgs {
+			pkgPath := "file:" + filepath.Join(localPath, "packages", pkg)
+			packageJSONDepStr += fmt.Sprintf(`"@temporalio/%v": %q,`, pkg, pkgPath)
+			packageJSONDepStr += "\n    "
+		}
+	} else {
+		packageJSONDepStr = `"temporalio": "` + strings.TrimPrefix(options.Version, "v") + "\",\n    "
+	}
+	packageJSON := `{
```

Comment:
My personal preference would be to have Go struct that's serialized as JSON but it's not like this is a regression from what we previously had, so I'm fine keeping this way.

Reply:
I think it's a bit much to want to model package json as a type-safe struct in Go just as I think it's a bit much to break this little thing out into a bunch of files. YAGNI, KISS, pick your adjective. This is a couple-hundred line file (without tests) to setup a TS project so I can move on.

Reply:
Mostly my argument is I just can't see how a template is any more effort than this (it strikes me as less). I agree a struct does seem like more effort.

But, whatever, it's not terribly important and you seem to like this so works for me

Reply:
It's not that much more effort (it's a little more setting it up with `go:embed`, at which point I might make the template a string literal, at which point I end up with this anyways). It's just that it's arguably of not much benefit. That people have to learn the Go template language to contribute and then it was still not as featureful as the other ones (I had to do extra work to make it work with relative paths from CLI and it was not written with cross-platform in mind because the JSON wasn't escaped).